src/QuadForm/indefiniteLLL.jl

```diff
@@ -0,0 +1,357 @@
+
+Extracts components of $M$ with regards to the binary representation of $x$ and $y$. 
+The location of the one entries give the position of the entries which should be extracted.
+For example if $x$ = $y$ = 13 = (1101)_2 then M[[1,3,4], [1,3,4]] is returned.
+
+$Example$
+==========
+julia> vecextract(ZZ[1 2 3; 4 5 6; 7 8 9],6,3)
+[4 5]
+[7 8]
+=#
+function _vecextract(M::MatElem,x::Int64,y::Int64)
+  x_bin = digits(x,base = 2)
+  y_bin = digits(y, base = 2)
+
+  list_x = [i for i = 1:length(x_bin) if x_bin[i] == 1]
+  list_y = [i for i = 1:length(y_bin) if y_bin[i] == 1]
+
+  return M[list_x , list_y]
+
+end
+
+function _vecextract(M::MatElem, x::Union{Int64,Vector{Int64}},y::Int64)
+  y_bin = digits(y, base = 2)
+  list_y = [i for i = 1:length(y_bin) if y_bin[i] == 1]
+
+  return M[x , list_y]
+
+end
+
+################################################################################
+#                           linear algebra
+################################################################################
+
+#=
+    _mathnf(A::MatElem) -> MatElem, MatElem
+
+Given a rectangular matrix $A$ of dimension $nxm$. Computes the Hessian matrix $H$ of dimension $nxm$ 
+and the unimodular transformation matrix $U$ such that $AU$ = $H$.
+=#
+function _mathnf(A::MatElem)
+  H, U = hnf_with_transform(reverse_cols(transpose(A)))
+  H = reverse_rows(reverse_cols(transpose(H)))
+  U = reverse_cols(transpose(U))
+
+  return H, U
+end
+
+#=
+    _complete_to_basis(v::MatElem; redflag::Bool = false) -> MatElem
+
+Given a rectangular matrix $nxm$ with $n$ != $m$ and redflag = 0.
+Computes a unimodular matrix with the last column equal to the last column of $v$.
+If redflag = 1, it LLL-reduce the $n$-$m$ first columns if $n$ > $m$.
+=#
+function _complete_to_basis(v::MatElem, redflag::Bool = false)
+  
+  n = nrows(v) #Number of rows
+  m = ncols(v) #Number of cols
+
+  if(n == m) #return nxn matrices
+    return v
+  end
+
+  U = inv(transpose(_mathnf(transpose(v))[2]))
+
+  if(n == 1 || redflag == false)
+    return U
+  end
+
+  re = U[:,1:n-m]
+  re = transpose(lll_with_transform(transpose(re))[2])
+
+  l = diagonal_matrix(re,one(parent(v[1:m,1:m])))
+  re = U*l
+
+  return re
+end
+
+#=
+    _ker_mod_p(M::MatElem,p) -> Int, MatElem
+
+Computes the kernel of the given matrix $M$ mod $p$.
+It returns [$rank$,$U$], where $rank$ = dim (ker M mod p) and $U$ in $GL_n$(Z),
+The first $rank$ columns of $U$ span the kernel.; check::Bool = false
+=#
+function _ker_mod_p(M::MatElem,p)
+  R = parent(M[1,1])
+  rk, k = kernel(change_base_ring(ResidueRing(R,p),M))
+  U = _complete_to_basis(lift(k[:,1:rk]))
+  reverse_cols!(U)
+
+  return rk, U
+end
+
+################################################################################
+#                           Quadratic Equations
+################################################################################
+
+#=
+    _quad_form_solve_triv(G::MatElem{fmpz}; base::Bool = false) -> MatElem{fmpz}
```

Comment:
Here your function is supposed to return 3 outputs, so you have to make it clear. However, since those internal functions and non-official documentation, as for the previous functions, you don't need this first line, only the texts explaining the function is relevant.